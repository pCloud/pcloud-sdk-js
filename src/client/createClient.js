/* @flow */

import invariant from "invariant";
import deepAssign from "deep-assign";
import ApiMethod from "../api/ApiMethod";
import type { ClientApiMethodOptions, ClientMethod, ClientType } from "./types";
import type { ApiResult } from "../api/types";
import * as methods from "./methods";
import { isAuthMethod } from "../utils";

const defaultApiServer = "api.pcloud.com";

export default function createClient(token: string, type: ClientType = "oauth", useProxy: boolean = false) {
	invariant(["oauth", "pcloud"].indexOf(type) !== -1, "`type` must be either `oauth` or `pcloud`.");

	if (type === "oauth") {
		invariant(typeof token === "string", "`token` is required.");
		invariant(token.length, "`token` is required.");
	}

	// Local Params
	// apiServer, token, type
	let apiServer = defaultApiServer;

	function initialOptions(method: string) {
		let options = { apiServer: apiServer, params: {} };

		if (isAuthMethod(method) && token) {
			options.params["oauth" === type ? "access_token" : "auth"] = token;
		}

		return options;
	}

	function api(method: string, options?: ClientApiMethodOptions = {}): Promise<ApiResult> {
		let mergeOptions = deepAssign({}, initialOptions(method), options);

		return ApiMethod(method, mergeOptions).catch(error => {
			if (error.result === 500 && apiServer !== defaultApiServer) {
				// reset API server
				apiServer = defaultApiServer;

				// retry
				return api(method, options);
			} else {
				return Promise.reject(error);
			}
		});
	}

	function setupProxy() {
		return api("getapiserver", {}).then((response: any) => {
			return (apiServer = response.api[0]);
		});
	}

	function setToken(newToken: string) {
		token = newToken;
	}

	// client api for end users
	let client: any = { api, setupProxy };

	let pcloudMethod;

	for (let method in methods) {
		if (methods.hasOwnProperty(method)) {
			let baseMethod: ClientMethod = methods[method];

			pcloudMethod = baseMethod({ client, setToken, type }, type);

			if (typeof pcloudMethod === "function") {
				client[method] = pcloudMethod;
			}
		}
	}

	if (useProxy) {
		setupProxy();
	}

	return client;
}
