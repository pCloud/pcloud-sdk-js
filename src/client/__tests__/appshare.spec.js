jest.mock("../../api/ApiMethod");
import apiMethod, { on, one, success, error } from "../../api/ApiMethod";

var shareSpy = jest.fn();
var cannotShareSpy = jest.fn();

on((method, { params: { userid } }) => method === "sharefolder" && userid === 1, success({}), shareSpy);
one(
	(method, { params: { userid } }) => method === "sharefolder" && userid === 2,
	error(2009, "File not found."),
	cannotShareSpy
);

import createClient from "../createClient";

describe("appshare", () => {
	it("sends correct data for appshare", async () => {
		const { appshare } = createClient("testauth", "pcloud", false);
		const response = await appshare(100, 1, "clientid");

		expect(shareSpy).toHaveBeenCalledTimes(1);
		expect(response).toBe(true);

		expect(apiMethod.mock.calls.length).toBe(1);
		expect(apiMethod.mock.calls[0][1].params).toEqual({
			auth: "testauth",
			client_id: "clientid",
			folderid: 100,
			permissions: 0,
			userid: 1
		});
	});

	it("throws for wrong token type", () => {
		const { appshare } = createClient("testauth", "oauth", false);

		expect(() => {
			appshare(100, 1, "clientid");
		}).toThrowErrorMatchingSnapshot();
	});

	it("throws for wrong token type", () => {
		const { appshare } = createClient("testauth", "pcloud", false);

		expect(() => {
			appshare(100, 1);
		}).toThrowError("clientid");
	});

	it("throws for wrong token type", () => {
		const { appshare } = createClient("testauth", "pcloud", false);

		expect(() => {
			appshare(100);
		}).toThrowError("userid");
	});
});
