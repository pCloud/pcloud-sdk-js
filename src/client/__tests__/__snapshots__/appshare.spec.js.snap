exports[`appshare throws for wrong token type 1`] = `"This method supports only clients of type \`pcloud\`."`;
